config.go

```diff
@@ -14,90 +14,104 @@ import (
 Config matchers for module interfaces.
 */
 
-func authProvider(modName string) (module.AuthProvider, error) {
-	modObj, err := module.GetInstance(modName)
-	if err != nil {
-		return nil, err
+// createModule is a helper function for config matchers that can create inline modules.
+func createModule(args []string) (module.Module, error) {
+	modName := args[0]
+	var instName string
+	if len(args) >= 2 {
+		instName = args[1]
+		if module.HasInstance(instName) {
+			return nil, fmt.Errorf("module instance named %s already exists", instName)
+		}
 	}
 
-	provider, ok := modObj.(module.AuthProvider)
-	if !ok {
-		return nil, fmt.Errorf("module %s doesn't implements auth. provider interface", modObj.Name())
+	newMod := module.Get(modName)
+	if newMod == nil {
+		return nil, fmt.Errorf("unknown module: %s", modName)
 	}
-	return provider, nil
-}
 
-func authDirective(m *config.Map, node *config.Node) (interface{}, error) {
-	if len(node.Args) != 1 {
-		return nil, m.MatchErr("expected 1 argument")
-	}
-	if len(node.Children) != 0 {
-		return nil, m.MatchErr("can't declare block here")
-	}
+	log.Debugln("module create", modName, instName, "(inline)")
 
-	modObj, err := authProvider(node.Args[0])
-	if err != nil {
-		return nil, m.MatchErr("%s", err.Error())
-	}
+	return newMod(modName, instName)
+}
 
-	return modObj, nil
+func initInlineModule(modObj module.Module, globals map[string]interface{}, node *config.Node) error {
+	// This is to ensure modules Init will see expected node layout if it breaks
+	// Map abstraction and works with map.Values.
+	//
+	// Expected: modName modArgs { ... }
+	// Actual: something modName modArgs { ... }
+	node.Name = node.Args[0]
+	node.Args = node.Args[1:]
+
+	log.Debugln("module init", modObj.Name(), modObj.InstanceName(), "(inline)")
+	return modObj.Init(config.NewMap(globals, node))
 }
 
-func storageBackend(modName string) (module.Storage, error) {
-	modObj, err := module.GetInstance(modName)
-	if err != nil {
-		return nil, err
+func authDirective(m *config.Map, node *config.Node) (interface{}, error) {
+	if len(node.Args) == 0 {
+		return nil, m.MatchErr("expected at least 1 argument")
 	}
 
-	backend, ok := modObj.(module.Storage)
-	if !ok {
-		return nil, fmt.Errorf("module %s doesn't implements storage interface", modObj.Name())
+	var modObj module.Module
+	var err error
+	if node.Children != nil {
+		modObj, err = createModule(node.Args)
+		if err != nil {
```

Comment:
We can check this once after the `if` since it's duplicated in both branches.

Maybe this should logic be extracted in a function, it's used several times.

Reply:
Done.